### PR TITLE
fix(testing): update jest batch mode

### PR DESCRIPTION
**Pull request by barbados-clemens**
jest go burrrr now

## Current Behavior

jest tests are slower in Nx than the jest CLI

(shout out @Coly010 for finding the config changes)

Note:
- requires using the `NX_BATCH_MODE=true` env var to be set.
- using `--output-style=stream` looks best/is easier to see what is going on (this should be the default for batch mode)
- ISSUE: if one project fails all projects are marked as failure. this is a bug that needs to be fixed in a follow up PR)

## Expected Behavior

test are the same speed as jest cli

## Related Issue(s)

TODO
- [ ] add tests for basic usage of `batchJest`
Fixes #

### CODE REVIEW

File: packages/jest/src/executors/jest/jest.impl.ts [hunk 1]
```diff
@@ -150,24 +150,36 @@ export async function batchJest(
   overrides: JestExecutorOptions,
   context: ExecutorContext
 ): Promise<Record<string, { success: boolean; terminalOutput: string }>> {
+  console.time('batchJest');
```
Comment (Coly010) [inline packages/jest/src/executors/jest/jest.impl.ts:153]: We should remove these

File: packages/jest/src/executors/jest/jest.impl.ts [hunk 2]
```diff
@@ -150,24 +150,36 @@ export async function batchJest(
...
   const configPaths = taskGraph.roots.map((root) =>
     path.resolve(context.root, inputs[root].jestConfig)
   );
-
+  const selectedProjects = taskGraph.roots.map((task) => task.split(':')[0]);
+  console.time('readConfigs');
```
Comment (Coly010) [inline packages/jest/src/executors/jest/jest.impl.ts:158]: Here

File: packages/jest/src/executors/jest/jest.impl.ts [hunk 3]
```diff
@@ -150,24 +150,36 @@ export async function batchJest(
...
+  const parsedConfigs = await jestConfigParser(overrides, context, true);
+
+  console.timeEnd('readConfigs');
```
Comment (Coly010) [inline packages/jest/src/executors/jest/jest.impl.ts:161]: Here

File: packages/jest/src/executors/jest/jest.impl.ts [hunk 4]
```diff
@@ -150,24 +150,36 @@ export async function batchJest(
...
+  // const configs = await Promise.all(
+  //   configPaths.map(
+  //     async (path) => await readConfig({ $0: '', _: undefined }, path)
+  //   )
+  // );
+  console.time('runCLI');
```
Comment (Coly010) [inline packages/jest/src/executors/jest/jest.impl.ts:167]: Here

File: packages/jest/src/executors/jest/jest.impl.ts [hunk 5]
```diff
@@ -150,24 +150,36 @@ export async function batchJest(
...
-    await jestConfigParser(overrides, context, true),
-    [...configPaths]
+    {
+      ...parsedConfigs,
+      selectProjects: selectedProjects,
+    },
+    [workspaceRoot]
   );
-
+  console.timeEnd('runCLI');
```
Comment (Coly010) [inline packages/jest/src/executors/jest/jest.impl.ts:175]: Here

File: packages/jest/src/executors/jest/jest.impl.ts [hunk 6]
```diff
@@ -150,24 +150,36 @@ export async function batchJest(
...
   const jestTaskExecutionResults: Record<
     string,
     { success: boolean; terminalOutput: string }
   > = {};
 
-  const configs = await Promise.all(
-    configPaths.map(async (path) => readConfig({ $0: '', _: undefined }, path))
-  );
-
+  console.time('getSummary');
```
Comment (Coly010) [inline packages/jest/src/executors/jest/jest.impl.ts:182]: Here

File: packages/jest/src/executors/jest/jest.impl.ts [hunk 7]
```diff
@@ -195,6 +208,8 @@ export async function batchJest(
       terminalOutput: resultOutput + '\n\r\n\r' + getSummary(aggregatedResults),
     };
   }
+  console.timeEnd('getSummary');
```
Comment (Coly010) [inline packages/jest/src/executors/jest/jest.impl.ts:211]: Here

File: packages/jest/src/executors/jest/jest.impl.ts [hunk 8]
```diff
@@ -195,6 +208,8 @@ export async function batchJest(
...
 
+  console.timeEnd('batchJest');
```
Comment (Coly010) [inline packages/jest/src/executors/jest/jest.impl.ts:213]: Here

**Review by Coly010 — CHANGES_REQUESTED** [review]
Left a bunch of comments just signalling where I saw console logs we should remove

File: packages/jest/src/executors/jest/jest.impl.ts [hunk 9]
```diff
@@ -153,21 +153,27 @@ export async function batchJest(
   const configPaths = taskGraph.roots.map((root) =>
     path.resolve(context.root, inputs[root].jestConfig)
   );
+  // TODO(caleb): this is depends on the name of the project to also match the 'displayName' in the jest.config.ts
+  //  this can also be an issue if someone has the same project name in a config.
+  //  i.e. two projects will be run for only 1 provided name
+  //  when trying to use readConfigs before runCLI ts throws error about root jest.config.ts.
+  const selectedProjects = taskGraph.roots.map((task) => task.split(':')[0]);
```
Comment (Cammisuli) [inline packages/jest/src/executors/jest/jest.impl.ts:160]: If you need to squeeze out a little more performance here, you can probably combine these into a single `for...of` loop and just push to separate arrays. 

It's not super important but it's potentially a nice optimization :D

**Review by Coly010 — APPROVED** [review]
LGTM!

**Review by Cammisuli — APPROVED** [review]

File: packages/jest/src/executors/jest/jest.impl.ts [hunk 10]
```diff
@@ -150,24 +156,50 @@ export async function batchJest(
...
+      console.warn(stripIndents`Could not find "displayName" in ${configPath}. 
+      Please add a "displayName" to the Jest Config File.
+      Skipping this project (${task.split(':')[0]})...`);
+      continue;
+    }
+
+    const displayName = fileContents
+      .match(displayNameValueRegex)
+      .map((value) => value.substring(1, value.length - 1))[0];
+    selectedProjects.push(displayName);
```
Comment (Coly010) [inline packages/jest/src/executors/jest/jest.impl.ts:184]: @barbados-clemens Here is the code to find the actual jest project 😄

Comment (barbados-clemens) [reply packages/jest/src/executors/jest/jest.impl.ts]: this should be fine for now, but there is a snowballs chance in hell that someone is using a var to reference the displayName. i.e. 

```ts
const myName = 'something-cool';

export default {
  displayName: myName
}
```
also more likely, 
```ts
export default {
  displayName: {
    name: 'something-cool',
    color: 'blue'
  }
}
```
in which case we'd error/warn about it missing when it was just configured differently. 

I think for an alpha it's fine, but we should make sure we handle this at some point. 

so maybe we should be more explicit in the error we are expecting the displayName to be a string for now?

File: packages/jest/src/executors/jest/jest.impl.ts [hunk 11]
```diff
@@ -150,24 +156,59 @@ export async function batchJest(
...
+
+    /* The display name in the jest.config.js is the correct project name jest
+     * uses to determine projects. It is usually the same as the Nx projectName
+     * but it can be changed. The safest method is to extract the displayName
+     * from the config file, but skip the project if it does not exist. */
+    const displayNameValueRegex = new RegExp(
+      /(['"]+.*['"])(?<=displayName+.*)/,
+      'g'
+    );
+    const fileContents = readFileSync(configPath, { encoding: 'utf-8' });
```
[the remark this answers is not in the document]
Comment (barbados-clemens) [reply packages/jest/src/executors/jest/jest.impl.ts:174]: agree. we should perf this and maybe read in parallel with the a promise. though idk how that will scale either with say 1000 projects trying to read in parallel.

[the remark this answers is not in the document]
Comment (Coly010) [reply packages/jest/src/executors/jest/jest.impl.ts:174]: I had the concern about potentially having too many processes in parallel, but we should definitely perf it

**Review by FrozenPandaz — APPROVED** [review]

File: packages/jest/src/executors/jest/jest.impl.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (bulldog98) [reply packages/jest/src/executors/jest/jest.impl.ts:174]: You could limit how many parallel calls are allowed

**Pull request merged**